core/summarize.py

```diff
@@ -36,57 +37,110 @@ def _get_incremental(loop_count, keys, data):
+    summary_by_service = summary_by_service.append(total, ignore_index=False)
+
+    storage_by_group.sort_index(inplace=True)
+    return ServiceSummary(summary_by_service, storage_by_group)
+
+
+def get_summary_data(config, service_data):
+    """Compute summary data for each month and each service"""
+
     # formula for compound growth: start_val * (1 + growth factor)^N
-    estimation_buffer = config.estimation_buffer * (1 + config.estimation_growth_factor) ** date_number
-    snapshot = service_data.loc[summary_date]
+    ebi = pd.Series(range(0, len(service_data.index)), index=service_data.index)
+    estimation_buffer = config.estimation_buffer * (1 + config.estimation_growth_factor) ** ebi
+    estimation_buffer = estimation_buffer.map(float)
+
     storage_units = config.storage_display_unit
     to_display = to_storage_display_unit(storage_units)
     to_gb = to_storage_display_unit('GB')
-    summary_df = pd.DataFrame()
+    summary_df = dict()
     for service_name, service_def in config.services.items():
-        service_snapshot = snapshot[service_name]
+        service_snapshot = service_data[service_name]
         compute = service_snapshot['Compute']
         data_storage = service_snapshot['Data Storage']['storage']
-        node_buffer = 0 if service_def.static_number else math.ceil(compute['VMs'] * float(estimation_buffer))
-        vms_suggested = math.ceil(compute['VMs'] + node_buffer)
-        vms_total = max(vms_suggested, service_def.min_nodes)
+        node_buffer = 0 if service_def.static_number else (compute['VMs'] * estimation_buffer).map(np.ceil)
+        vms_suggested = (compute['VMs'] + node_buffer).map(np.ceil)
+        vms_total = pd.DataFrame([
+            vms_suggested,
+            pd.Series([service_def.min_nodes] * len(vms_suggested), index=vms_suggested.index)
+        ]).max()
 
         storage_estimation_buffer = estimation_buffer
         if service_def.storage.override_estimation_buffer is not None:
-            storage_estimation_buffer = service_def.storage.override_estimation_buffer
+            storage_estimation_buffer = pd.Series(
+                [float(service_def.storage.override_estimation_buffer)] * len(service_data.index),
+                index=service_data.index
+            )
+        # this is True if 'service_def.min_nodes' is more than what is being suggested
+        vm_total_gt = vms_total > vms_suggested
+        data_storage_buffer = data_storage * storage_estimation_buffer
 
         if service_def.storage_scales_with_nodes:
-            data_storage_buffer = data_storage * float(storage_estimation_buffer)
+            # in this case total storage = storage * VM number
             data_storage_per_vm = data_storage + data_storage_buffer
             data_storage_total = data_storage_per_vm * vms_total
-        elif vms_total > vms_suggested:
+        elif True in set(vm_total_gt):
             # in this case 'service_def.min_nodes' is more than what is being suggested
```

Comment:
Just to make sure I'm following, this only gets hit if the current calculation is less than the minimum nodes?
And minimum node is just a field in the config file, right? How is that calculated?

Reply:
yes, that's correct. it's mostly used for things like riak (min 5 nodes) or couch (min 3 nodes).